### ISSUE TITLE
[$550] Filestack plugin for forums

### ISSUE THREAD

**Issue by jmgasper** [topcoder-platform/forums#28]
@atelomycterus - One additional plugin we'd like to create is a Filestack plugin, for managing attachments, uploaded by copilots and admins.  Interested to know your thoughts on this:

### Currently

Currently, Vanilla just saves uploads to the file system, which is less than ideal, as we'll need to handle many, many large uploads without worrying about running out of space, and we'll also need to virus scan everything.

### Community submission

The community app uses a Filestack file picker, which is backed by an Amazon S3 bucket, for challenge submissions.  We'd like to do something similar for Vanilla, replacing the existing upload.

### Expectation

If we turn on the Filestack plugin, uploads will be handled by Filestack.  All configuration should be exposed onto the plugin itself, in the management dashboard.

Ideally, the attachment functionality would remain as it does now in Vanilla, just a different file picker / upload.

### Code location

We'll use this repo for the code:  https://github.com/topcoder-platform/forums-filestack-plugin.  I've requested you be granted access - it's currently empty.  This will be a new plugin, since we want to keep it separate in the hopes that we may open-source it.

### Documentation

From Topcoder:

Here’s where we keep the submission page drop-in in community-app

https://github.com/topcoder-platform/community-app/blob/74907e6b0f93b2979a9db14606c6b11d3fb3e8bc/src/shared/components/SubmissionPage/FilestackFilePicker/index.jsx

Basically the way it works is that you configure Filestack to use an S3 bucket as storage. This can be a private bucket.
The uploader will send the file up to Filestack, do any Filestack post processing like AV scans, and put the file in the bucket for you. It then returns you the URL of that file in the bucket. That will allow you to save the URL as the reference to the file and fetch it via some other backend later

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30140390 has been created for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30140390 has been updated - it has been assigned to obog.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by atelomycterus** [comment]
@jmgasper  Filestack provides official PHP SDK. Plan to use it. 


> Community submission
> The community app uses a Filestack file picker, which is backed by an Amazon S3 bucket, for challenge submissions. We'd like to do something similar for Vanilla, replacing the existing upload.

I've looked at Vanilla source code. There is no separate Vanilla control for uploading a file out of the box.

**Vanilla Web Site** 
Uploading files is implemented with editor plugins(Advanced Editor, Rich Editor) in Vanilla. Both plugins are available in Open Source Community Vanilla by default.

Advanced Editor is an older editor that enables advanced editing of posts in several formats, including WYSIWYG, simple HTML, Markdown, and BBCode. Typically used by legacy customers, we recommend using our Rich Editor instead.
Details from https://success.vanillaforums.com/kb/articles/12-advanced-editor

Uploading an image:
![image](https://user-images.githubusercontent.com/10196632/92154418-7100b700-ee2e-11ea-97c5-247840b7fde4.png)


Uploading a file/attachment:
![image](https://user-images.githubusercontent.com/10196632/92154551-a3121900-ee2e-11ea-92d2-9c07269cd18f.png)

Paths are saved in the  Media table
![image](https://user-images.githubusercontent.com/10196632/92162913-90521100-ee3b-11ea-99ba-611e666afacb.png)

**Admin site**

Vanilla developers use php directly to generate a html form with file upload control in admin area.
See fileUpload/uploadImage from 'package/library/core/class.form.php' 

---------

> If we turn on the Filestack plugin, uploads will be handled by Filestack. All configuration should be exposed onto the plugin itself, in the management dashboard.
> 
> Ideally, the attachment functionality would remain as it does now in Vanilla, just a different file picker / upload.

1. Should we integrate with Vanilla editors? If yes, which Vanilla editor (Advanced Editor, Rich Editor) ?   

2. Should we develop a basic control (with Filestack file picker, Javascript) which will be used in a custom form/view?

**Comment by jmgasper** [comment]
1.  Let's target the Rich Editor
2.  If that's easiest, yes, let's do that, thanks.

**Comment by atelomycterus** [comment]
@jmgasper Implemented with an editor. Need you feedback. 
Unfortantelly, I don't have permissions to push it to 'forums-filestack-plugin'.

### PHP Configuration
PHP file upload size is set to maximum 2MB file on the server by default, 
but you can increase or decrease the maximum size of file upload using the PHP configuration file (php.ini)
To increaes file upload size in PHP, you need to modify the upload_max_filesize and post_max_size variable’s in your php.ini file.

upload_max_filesize = 10M
post_max_size = 10M
max_file_uploads = 25

Accepted Measurements: Megabyte - M, Gigabyte - G

### Vanilla Configuration
Vanilla has some config parameters for file handling:
$Configuration['Garden']['Upload']['MaxFileSize'] = '50M';
$Configuration['Garden']['Upload']['AllowedFileExtensions'] = [
    'txt', 'jpg', 'jpeg', 'gif', 'png', 'bmp', 'tiff', 'ico', 'zip', 'gz', 'tar.gz', 'tgz', 'psd', 'ai', 'pdf', 'doc', 'xls', 'ppt', 'docx', 'xlsx', 'pptx', 'log', 'rar', '7z'
];

### Filestack Plugin settings
The plugin re-use vanilla config parameters for file handling. If the 'MaxFileSize' are not set then values of 'upload_max_filesize' is used.

![image](https://user-images.githubusercontent.com/10196632/92306612-20599d00-ef99-11ea-84fd-af39c9fc6668.png)

### Editor plugins

**Advanced Editor**
I've integrated this plugin with Filestack using the 'editorUpload_create' hook.

Video: https://monosnap.com/file/IEmAkZaHwb8jtQAMbR3E8yYiOtRuqb
I added attachments to a post, added files which a part of post body. All files were uploaded to Filestack.
Then I disabled the plugin. All new files were uploaded to '/upload'.
The previous links of files were valid and available from CDN.  

**Rich Editor**
The front-end of Rich editor is written Typescript and uses REST API (/api/v2/media) to upload files.
At first, it looked easy to add FilestackApiController and replace the implementation of MediaApiController, but it didn't work.
Vanilla uses an dependency injection container to configure base classes, factories, interfaces including REST API.
It's not possible to update/reload/destroy a bean after adding it into the container.  
So not as powerful as Java Spring for example.

If add a new rest API (/api/v2/filestack), then a lot of changes are required. The plugin depends 
on core Vanilla Typescript library where rest API path hardcoded. There will be too much changes in the plugin.

Another option is to check if route can be used with a plugin ('/media => /filestack). Let me know if it's required.

**Comment by atelomycterus** [comment]
@jmgasper  Keep you updated.

The Advanced editor is very tightly integrated with posts and comments.  The server side 'editorUpload_create' hook is used 
to upload files to Filestack.

Keep in mind

>  'we'll need to handle many, many large uploads without worrying about running out of space, and we'll also need to virus scan everything.'

So it'd be better to integrate with Javascript Web picker and upload files from a client browser to Filestack.

### Using Javascript Web picker with a custom page

In short, the plugin creates a page ("/filestack") with the Javascript Web picker control. The user can upload files directly to Filestake. If the plugin is disabled then the page is unregistered.
Javascript Web picker uses settings of Filestack Plugin including API Key, Vanilla max file size and supported file extensions. 
This example demonstrates that we can add any functionality with a custom Vanilla page. 
For example, upon completion of the upload, call the Vanilla REST API and perform some actions.
 
Video: https://monosnap.com/file/VhHZfajWKSR74zDtbRwGbVp46zVavh

The custom page with a web picker:
![image](https://user-images.githubusercontent.com/10196632/92312935-76950300-efce-11ea-9e73-63cefbcc030f.png)

![image](https://user-images.githubusercontent.com/10196632/92312916-39307580-efce-11ea-9f09-7ad691b46cd3.png)

Upload file and show the file link:
![image](https://user-images.githubusercontent.com/10196632/92312909-29189600-efce-11ea-9722-51b704a9cdff.png)

**Comment by jmgasper** [comment]
@atelomycterus - Ok, sounds good.

Can we update the CI/CD to install this plugin please?  That way I can test in dev and show it to Topcoder to get feedback, thanks.  I'll bump the payment here for the ask.

**Comment by atelomycterus** [comment]
@jmgasper  I don't have permissions to "https://github.com/topcoder-platform/forums-filestack-plugin". 
I can push the code to 'https://github.com/topcoder-platform/forums-plugins'.  Later we can add it in the forums-filestack-plugin repo.   Are you ok with that?

**Comment by jmgasper** [comment]
@atelomycterus - Let's wait until we can get the permissions set.  Unfortunately, I don't have the correct permissions to forums-filestack-plugin either 🤦

**Comment by jmgasper** [comment]
@atelomycterus - Access is now fixed.

**Comment by atelomycterus** [comment]
@jmgasper Yes. Thanks.  So we use a private repo and need to authorize our CI/CD to checkout it.
It can be configured different ways (https://enquizit.com/2019/06/05/how-to-authorize-your-ci-cd-to-checkout-additional-repo/). 

Do you want to use a machine token or ssh key pair?

**Comment by atelomycterus** [comment]
@jmgasper  Created PR-https://github.com/topcoder-platform/forums/pull/31. Need to create a machine token and set the 'CI_DEPLOY_TOKEN' env.  This token will be used to clone a private repo.

1) Enable the Filestack plugin and set APIKEY
2) Use Advanced Editor.  Set 'Wysiwyg' in  Posting page and click on 'Save'. 'Text'/'Rich' is used in Rich Editor.

![image](https://user-images.githubusercontent.com/10196632/92374250-24f09380-f108-11ea-95f6-b63b6ca33e73.png)

### Using JS Web Picker
Go to 'http://127.0.0.1/filestack'

### Using Advanced Editor
Add/Edit a discussion/comment and upload  files.

**Comment by jmgasper** [comment]
@atelomycterus - Thanks for that.  I'm asking Topcoder if I can make the plugin repo public.  That'd be the easiest tweak, so hopefully they go for that.

**Comment by atelomycterus** [comment]
@jmgasper Ok. Let me know if  the repo will be public. I'll remove 'CI_DEPLOY_TOKEN from the code.

**Comment by jmgasper** [comment]
@atelomycterus - Deployed, looks great, thanks!

**Comment by jmgasper** [comment]
Payment task has been updated: https://software.topcoder.com/review/actions/ViewProjectDetails?pid=30140390<br/><br/>```This is an automated message for ghostar via Topcoder X```
